File: bridgebase_sdk.py
"""
BridgeBase Python SDK
The Trust Layer for the Agentic Economy

Usage:
    from bridgebase_sdk import BridgeBaseClient

    client = BridgeBaseClient(gateway_url="https://pqc-gateway-production.up.railway.app")

    # Register a new agent (run once, save private key)
    result = client.register(agent_id="my-agent-001")
    private_key = result["private_key"]  # SAVE THIS

    # Authenticate (run every session)
    token = client.authenticate(agent_id="my-agent-001", private_key=private_key)

    # Gate a transaction
    cleared = client.validate_token(agent_id="my-agent-001", session_token=token)
"""

import httpx
from typing import Optional


class BridgeBaseError(Exception):
    """Raised when the gateway returns an error."""
    pass


class BridgeBaseClient:
    """
    Synchronous client for the BridgeBase PQC Gateway.
    No dependencies beyond httpx.
    """

    def __init__(
        self,
        gateway_url: str = "https://pqc-gateway-production.up.railway.app",
        timeout: int = 30,
    ):
        self.gateway_url = gateway_url.rstrip("/")
        self.timeout = timeout

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _post(self, path: str, body: dict) -> dict:
        url = f"{self.gateway_url}{path}"
        try:
            r = httpx.post(url, json=body, timeout=self.timeout)
        except httpx.ConnectError:
            raise BridgeBaseError(f"Cannot connect to gateway at {self.gateway_url}")
        if not r.is_success:
            detail = r.json().get("detail", r.text) if r.content else r.status_code
            raise BridgeBaseError(f"Gateway error on {path}: {detail}")
        return r.json()

    def _get(self, path: str) -> dict:
        url = f"{self.gateway_url}{path}"
        try:
            r = httpx.get(url, timeout=self.timeout)
        except httpx.ConnectError:
            raise BridgeBaseError(f"Cannot connect to gateway at {self.gateway_url}")
        if not r.is_success:
            detail = r.json().get("detail", r.text) if r.content else r.status_code
            raise BridgeBaseError(f"Gateway error on {path}: {detail}")
        return r.json()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def health(self) -> dict:
        """Check if the gateway is online."""
        return self._get("/health")

    def register(self, agent_id: str, metadata: Optional[dict] = None) -> dict:
        """
        Register a new agent. Returns private key — save it, never stored again.

        Returns:
            {
                "agent_id": str,
                "public_key": str,
                "private_key": str,   # base64 ML-KEM-768 private key
                "algorithm": str,
            }
        """
        body = {"agent_id": agent_id, "metadata": metadata or {}}
        data = self._post("/register", body)
        return {
            "agent_id": data["agent_id"],
            "public_key": data["public_key"],
            "private_key": data["private_key_plaintext"],
            "algorithm": data["algorithm"],
        }

    def authenticate(self, agent_id: str, private_key: str) -> str:
        """
        Full 3-step PQC handshake. Returns session token on success.

        Steps handled automatically:
            1. Request challenge from gateway
            2. Decapsulate ciphertext using private key
            3. Verify with gateway → receive session token

        Args:
            agent_id:    Your registered agent ID
            private_key: Base64 ML-KEM-768 private key from register()

        Returns:
            session_token (str) — pass this to validate_token() before transactions
        """
        # Step 1: Challenge
        ch = self._post("/challenge", {"agent_id": agent_id})

        # Step 2: Solve (decapsulate via gateway dev endpoint)
        solved = self._post("/dev/solve", {
            "private_key_b64": private_key,
            "ciphertext_b64": ch["ciphertext"],
        })

        # Step 3: Verify
        verified = self._post("/verify", {
            "agent_id": agent_id,
            "challenge_id": ch["challenge_id"],
            "shared_secret_b64": solved["shared_secret_b64"],
        })

        return verified["session_token"]

    def validate_token(self, agent_id: str, session_token: str) -> bool:
        """
        Check if a session token is valid before executing a transaction.

        Returns:
            True if cleared, raises BridgeBaseError if blocked.
        """
        result = self._post("/validate-token", {
            "agent_id": agent_id,
            "session_token": session_token,
        })
        return result["cleared"]

    def get_passport(self, agent_id: str) -> dict:
        """Fetch an agent's public passport (public key + reputation)."""
        return self._get(f"/passport/{agent_id}")

    def list_agents(self) -> list:
        """List all registered agent IDs."""
        return self._get("/agents")["agents"]

    def reputation(self, agent_id: str) -> int:
        """Get an agent's current reputation score."""
        passport = self.get_passport(agent_id)
        return passport["reputation_score"]

File: bridgebase-sdk.js
/**
 * BridgeBase JavaScript SDK
 * The Trust Layer for the Agentic Economy
 * 
 * Works in: Node.js, ElizaOS, Virtuals, browser
 * No dependencies — uses built-in fetch only
 * 
 * Usage:
 *   const { BridgeBaseClient } = require('./bridgebase-sdk')
 *   const client = new BridgeBaseClient()
 *   const token = await client.authenticate('my-agent', privateKey)
 *   const cleared = await client.validateToken('my-agent', token)
 */

'use strict'

class BridgeBaseError extends Error {
  constructor(message, statusCode = null) {
    super(message)
    this.name = 'BridgeBaseError'
    this.statusCode = statusCode
  }
}

class BridgeBaseClient {
  /**
   * @param {object} options
   * @param {string} options.gatewayUrl - BridgeBase gateway URL
   * @param {number} options.timeout    - Request timeout in ms (default 30000)
   */
  constructor(options = {}) {
    this.gatewayUrl = (options.gatewayUrl || 'https://pqc-gateway-production.up.railway.app').replace(/\/$/, '')
    this.timeout = options.timeout || 30000
  }

  // -------------------------------------------------------------------------
  // Internal
  // -------------------------------------------------------------------------

  async _post(path, body) {
    const url = `${this.gatewayUrl}${path}`
    let res
    try {
      res = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(this.timeout),
      })
    } catch (err) {
      throw new BridgeBaseError(`Cannot connect to gateway: ${err.message}`)
    }

    const data = await res.json().catch(() => ({}))

    if (!res.ok) {
      const detail = data.detail || data.message || res.statusText
      throw new BridgeBaseError(`Gateway error on ${path}: ${detail}`, res.status)
    }

    return data
  }

  async _get(path) {
    const url = `${this.gatewayUrl}${path}`
    let res
    try {
      res = await fetch(url, {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout),
      })
    } catch (err) {
      throw new BridgeBaseError(`Cannot connect to gateway: ${err.message}`)
    }

    const data = await res.json().catch(() => ({}))

    if (!res.ok) {
      const detail = data.detail || data.message || res.statusText
      throw new BridgeBaseError(`Gateway error on ${path}: ${detail}`, res.status)
    }

    return data
  }

  // -------------------------------------------------------------------------
  // Public API
  // -------------------------------------------------------------------------

  /**
   * Check gateway health
   * @returns {{ status: string, algorithm: string }}
   */
  async health() {
    return this._get('/health')
  }

  /**
   * Register a new agent — run once, save the private key
   * @param {string} agentId
   * @param {object} metadata
   * @returns {{ agentId, publicKey, privateKey, algorithm }}
   */
  async register(agentId, metadata = {}) {
    const data = await this._post('/register', {
      agent_id: agentId,
      metadata,
    })
    return {
      agentId: data.agent_id,
      publicKey: data.public_key,
      privateKey: data.private_key_plaintext,
      algorithm: data.algorithm,
    }
  }

  /**
   * Full 3-step PQC handshake — returns session token
   * Handles challenge + solve + verify automatically
   * @param {string} agentId
   * @param {string} privateKey - base64 ML-KEM-768 private key from register()
   * @returns {string} session token
   */
  async authenticate(agentId, privateKey) {
    // Step 1: Request challenge
    const challenge = await this._post('/challenge', { agent_id: agentId })

    // Step 2: Solve (decapsulate via gateway)
    const solved = await this._post('/dev/solve', {
      private_key_b64: privateKey,
      ciphertext_b64: challenge.ciphertext,
    })

    // Step 3: Verify → get session token
    const verified = await this._post('/verify', {
      agent_id: agentId,
      challenge_id: challenge.challenge_id,
      shared_secret_b64: solved.shared_secret_b64,
    })

    return verified.session_token
  }

  /**
   * Validate session token before a transaction
   * @param {string} agentId
   * @param {string} sessionToken
   * @returns {boolean} true if cleared
   * @throws {BridgeBaseError} if blocked (403)
   */
  async validateToken(agentId, sessionToken) {
    const result = await this._post('/validate-token', {
      agent_id: agentId,
      session_token: sessionToken,
    })
    return result.cleared === true
  }

  /**
   * Fetch agent public passport
   * @param {string} agentId
   * @returns {{ agentId, publicKey, reputationScore, registeredAt }}
   */
  async getPassport(agentId) {
    const data = await this._get(`/passport/${agentId}`)
    return {
      agentId: data.agent_id,
      publicKey: data.public_key,
      reputationScore: data.reputation_score,
      registeredAt: data.registered_at,
      metadata: data.metadata,
    }
  }

  /**
   * List all registered agent IDs
   * @returns {string[]}
   */
  async listAgents() {
    const data = await this._get('/agents')
    return data.agents
  }

  /**
   * Get agent reputation score
   * @param {string} agentId
   * @returns {number}
   */
  async reputation(agentId) {
    const passport = await this.getPassport(agentId)
    return passport.reputationScore
  }
}

module.exports = { BridgeBaseClient, BridgeBaseError }
